schema/types.go

```diff
@@ -15,77 +14,241 @@ const (
-			err = t.Set(v)
-		case TypeInetArray:
-			t := &cqtypes.InetArray{}
-			err = t.Set(v)
-		case TypeCIDR:
-			t := &cqtypes.CIDR{}
-			err = t.Set(v)
-		case TypeCIDRArray:
-			t := &cqtypes.CIDRArray{}
-			err = t.Set(v)
-		case TypeMacAddr:
-			t := &cqtypes.Macaddr{}
-			err = t.Set(v)
-		case TypeMacAddrArray:
-			t := &cqtypes.MacaddrArray{}
-			err = t.Set(v)
-		default:
-			return nil, fmt.Errorf("unsupported type %s", table.Columns[i].Type)
+	}
+	return false
+}
+
+const (
+	TypeInvalid ValueType = iota
+	TypeBool
+	TypeInt
+	TypeFloat
+	TypeUUID
+	TypeString
+	TypeByteArray
+	TypeStringArray
+	TypeIntArray
+	TypeTimestamp
+	TypeJSON
+	TypeUUIDArray
+	TypeInet
+	TypeInetArray
+	TypeCIDR
+	TypeCIDRArray
+	TypeMacAddr
+	TypeMacAddrArray
+	TypeTimeIntervalDeprecated
+	TypeEnd
+)
+
+func (r *ValueType) UnmarshalJSON(data []byte) (err error) {
+	var valueType int
+	if err := json.Unmarshal(data, &valueType); err != nil {
+		return err
+	}
+	if ValueType(valueType) <= TypeInvalid || ValueType(valueType) >= TypeEnd {
+		*r = TypeInvalid
+	} else {
+		*r = ValueType(valueType)
+	}
+	return nil
+}
+
+func (r ValueType) String() string {
+	switch r {
+	case TypeBool:
+		return "TypeBool"
+	case TypeInt:
+		return "TypeInt"
+	case TypeFloat:
+		return "TypeFloat"
+	case TypeUUID:
+		return "TypeUUID"
+	case TypeString:
+		return "TypeString"
+	case TypeJSON:
+		return "TypeJSON"
+	case TypeIntArray:
+		return "TypeIntArray"
+	case TypeStringArray:
+		return "TypeStringArray"
+	case TypeTimestamp:
+		return "TypeTimestamp"
+	case TypeByteArray:
+		return "TypeByteArray"
+	case TypeUUIDArray:
+		return "TypeUUIDArray"
+	case TypeInetArray:
+		return "TypeInetArray"
+	case TypeInet:
+		return "TypeInet"
+	case TypeMacAddrArray:
+		return "TypeMacAddrArray"
+	case TypeMacAddr:
+		return "TypeMacAddr"
+	case TypeCIDRArray:
+		return "TypeCIDRArray"
+	case TypeCIDR:
+		return "TypeCIDR"
+	case TypeInvalid:
+		return "TypeInvalid"
+	default:
+		return fmt.Sprintf("Unknown(%d)", r)
+	}
+}
+
+type CQType interface {
+	Set(v interface{}) error
+	Get() interface{}
+	String() string
+	Equal(CQType) bool
+	Type() ValueType
+}
+
+type CQTypes []CQType
+
+type cqTypeWrapper struct {
+	Type  ValueType `json:"type"`
+	Value CQType    `json:"value"`
+}
+
+func (c CQTypes) MarshalJSON() ([]byte, error) {
+	res := make([]*cqTypeWrapper, len(c))
+	for i, v := range c {
+		res[i] = &cqTypeWrapper{
+			Type:  v.Type(),
+			Value: v,
 		}
-		if err != nil {
-			return nil, fmt.Errorf("failed to convert value %v to type %s: %w", v, table.Columns[i].Type, err)
+	}
+	return json.Marshal(res)
+}
+
+func (c *CQTypes) UnmarshalJSON(b []byte) error {
+	var res []map[string]json.RawMessage
+	if err := json.Unmarshal(b, &res); err != nil {
+		return err
+	}
+	cqTypes := make(CQTypes, 0, len(res))
+	for i := range res {
+		var t ValueType
+		if err := json.Unmarshal(res[i]["type"], &t); err != nil {
+			return err
+		}
+		if t <= TypeInvalid || t >= TypeEnd {
```

Comment:
Interesting technique to check `t >= TypeEnd`, but I like it 😄

Reply:
it's very C (`MAX_TYPE` would be the C counterpart)